File: src/services/SeedInfo/infoHandler/InfoProviders/Alchemy/alchemy.cpp
#include "../../../noita_random/src/noita_random.cpp"

class MaterialPicker
{
public:
  vector<string> LIQUIDS = {
      "acid", "alcohol", "blood",
      "blood_fungi", "blood_worm", "cement",
      "lava", "magic_liquid_berserk", "magic_liquid_charm",
      "magic_liquid_faster_levitation", "magic_liquid_faster_levitation_and_movement", "magic_liquid_invisibility",
      "magic_liquid_mana_regeneration", "magic_liquid_movement_faster", "magic_liquid_protection_all",
      "magic_liquid_teleportation", "magic_liquid_unstable_polymorph", "magic_liquid_unstable_teleportation",
      "magic_liquid_worm_attractor", "material_confusion", "mud",
      "oil", "poison", "radioactive_liquid_yellow",
      "swamp", "urine", "water",
      "water_ice", "water_swamp", "magic_liquid_random_polymorph"};

  vector<string> ALCHEMY = {
      "bone_box2d", "brass", "coal",
      "copper", "diamond", "fungi",
      "gold", "grass", "gunpowder",
      "gunpowder_explosive", "rotten_meat", "sand_petrify",
      "silver", "slime", "snow_b2",
      "soil", "wax", "honey"};

  NollaPrng *PRNG;

  vector<string> Materials = {};

  MaterialPicker(NollaPrng *prng, uint worldSeed)
  {
    PRNG = prng;
    PickMaterials(LIQUIDS, 3);
    PickMaterials(ALCHEMY, 1);
    ShuffleList(worldSeed);
    PRNG->Next();
    PRNG->Next();
  }

  void PickMaterials(vector<string> source, int count)
  {
    int counter = 0;
    int failed = 0;
    while (counter < count && failed < 99999)
    {
      int rand = (int)(PRNG->Next() * (double)source.size());
      string picked = source[rand];
      if (std::find(Materials.begin(), Materials.end(), picked) == Materials.end())
      {
        Materials.push_back(picked);
        counter++;
      }
      else
      {
        failed++;
      }
    }
    return;
  }

  void ShuffleList(uint worldSeed)
  {
    NollaPrng *prng = new NollaPrng((worldSeed >> 1) + 12534);

    for (int i = Materials.size() - 1; i >= 0; i--)
    {
      int rand = prng->Next() * (double)(i + 1);
      string tmp = Materials[i];
      Materials[i] = Materials[rand];
      Materials[rand] = tmp;
    }
    delete prng;
  }
};

static string PickForSeed(uint world_seed)
{
  NollaPrng *prng = new NollaPrng((world_seed * 0.17127000) + 1323.59030000);
  // Preheat random!
  for (int i = 0; i < 5; i++)
  {
    prng->Next();
  }

  MaterialPicker *lc = new MaterialPicker(prng, world_seed);
  string slc = "lc:" + lc->Materials[0] + "," + lc->Materials[1] + "," + lc->Materials[2]; // + "," + lc->Materials[3];
  MaterialPicker *ap = new MaterialPicker(prng, world_seed);
  string sap = "ap:" + ap->Materials[0] + "," + ap->Materials[1] + "," + ap->Materials[2]; // + "," + ap->Materials[3];

  delete prng;
  delete lc;
  delete ap;

  return slc + ";" + sap;
}

#include <emscripten/bind.h>

using namespace emscripten;

EMSCRIPTEN_BINDINGS(noitool_alchemy)
{
  emscripten::function<string>("PickForSeed", &PickForSeed);
}

File: src/services/SeedInfo/wasm/entries/alchemy.zig
const prng = @import("nolla_prng");
const mat = @import("materials");

const liquids = [_]u16{
    mat.acid,
    mat.alcohol,
    mat.blood,
    mat.blood_fungi,
    mat.blood_worm,
    mat.cement,
    mat.lava,
    mat.magic_liquid_berserk,
    mat.magic_liquid_charm,
    mat.magic_liquid_faster_levitation,
    mat.magic_liquid_faster_levitation_and_movement,
    mat.magic_liquid_invisibility,
    mat.magic_liquid_mana_regeneration,
    mat.magic_liquid_movement_faster,
    mat.magic_liquid_protection_all,
    mat.magic_liquid_teleportation,
    mat.magic_liquid_unstable_polymorph,
    mat.magic_liquid_unstable_teleportation,
    mat.magic_liquid_worm_attractor,
    mat.material_confusion,
    mat.mud,
    mat.oil,
    mat.poison,
    mat.radioactive_liquid,
    mat.swamp,
    mat.urine,
    mat.water,
    mat.water_ice,
    mat.water_swamp,
    mat.magic_liquid_random_polymorph,
};

const alchemy = [_]u16{
    mat.bone,
    mat.brass,
    mat.coal,
    mat.copper,
    mat.diamond,
    mat.fungi,
    mat.gold,
    mat.grass,
    mat.gunpowder,
    mat.gunpowder_explosive,
    mat.rotten_meat,
    mat.sand,
    mat.silver,
    mat.slime,
    mat.snow,
    mat.soil,
    mat.wax,
    mat.honey,
};

var result = [_]u16{0} ** 6;

fn contains(items: []const u16, len: usize, item: u16) bool {
    for (items[0..len]) |existing| {
        if (existing == item) {
            return true;
        }
    }
    return false;
}

fn pickMaterials(rng: *prng.NollaPrng, materials: *[4]u16, material_count: *usize, source: []const u16, count: usize) void {
    var counter: usize = 0;
    var failed: usize = 0;
    while (counter < count and failed < 99999) {
        const index: usize = @intFromFloat(rng.next() * @as(f64, @floatFromInt(source.len)));
        const picked = source[index];
        if (!contains(materials, material_count.*, picked)) {
            materials[material_count.*] = picked;
            material_count.* += 1;
            counter += 1;
        } else {
            failed += 1;
        }
    }
}

fn pickForOutput(rng: *prng.NollaPrng, world_seed: u32, output: []u16) void {
    var materials = [_]u16{0} ** 4;
    var material_count: usize = 0;
    pickMaterials(rng, &materials, &material_count, &liquids, 3);
    pickMaterials(rng, &materials, &material_count, &alchemy, 1);

    var shuffle_rng = prng.NollaPrng.init(@as(f64, @floatFromInt(world_seed >> 1)) + 12534.0);
    var i: i32 = @intCast(material_count - 1);
    while (i >= 0) : (i -= 1) {
        const limit = @as(f64, @floatFromInt(i + 1));
        const index: usize = @intFromFloat(shuffle_rng.next() * limit);
        const current: usize = @intCast(i);
        const tmp = materials[current];
        materials[current] = materials[index];
        materials[index] = tmp;
    }

    _ = rng.next();
    _ = rng.next();

    output[0] = materials[0];
    output[1] = materials[1];
    output[2] = materials[2];
}

export fn PickAlchemy(world_seed: u32) void {
    var rng = prng.NollaPrng.init((@as(f64, @floatFromInt(world_seed)) * 0.17127000) + 1323.59030000);
    for (0..5) |_| {
        _ = rng.next();
    }

    pickForOutput(&rng, world_seed, result[0..3]);
    pickForOutput(&rng, world_seed, result[3..6]);
}

export fn GetAlchemyMaterial(index: u32) u32 {
    if (index >= result.len) {
        return 0;
    }
    return result[index];
}
